godtools/App/Share/Data/TranslationsRepository/TranslationsRepository.swift

```diff
@@ -34,195 +34,123 @@ class TranslationsRepository {
-
-                let failedResponseHttpStatusCodes: [Int] = data.fileResponses.compactMap({$0.urlResponseObject?.httpStatusCode}).filter({$0 >= 400})
-                let failedToDownloadRelatedFile: Bool = failedResponseHttpStatusCodes.count > 0
+        let manifestParser: TranslationManifestParser = self.getManifestParser(manifestParserType: manifestParserType)
+        
+        return manifestParser.parse(manifestName: translation.manifestName).publisher
+            .map { manifest in
                 
-                if !failedToDownloadRelatedFile {
-                    
-                    let translationFilesDataModel = TranslationFilesDataModel(
-                        translationId: translationId,
-                        manifestFileName: data.manifestName,
-                        fileCacheLocations: data.fileResponses.compactMap({$0.fileCacheLocation})
-                    )
-                    
-                    return Just(translationFilesDataModel).setFailureType(to: Error.self)
-                        .eraseToAnyPublisher()
-                }
-                else {
-                    
-                    return self.downloadAndCacheTranslationZipFileFromRemote(translationId: translationId)
-                        .map { zipFileResponse in
-                            
-                            let translationFilesDataModel = TranslationFilesDataModel(
-                                translationId: translationId,
-                                manifestFileName: data.manifestName,
-                                fileCacheLocations: zipFileResponse.fileCacheLocations
-                            )
-                            
-                            return translationFilesDataModel
-                        }
-                        .eraseToAnyPublisher()
-                }
-            })
+                return TranslationManifestFileDataModel(manifest: manifest, translation: translation)
+            }
             .eraseToAnyPublisher()
     }
     
-    private func getTranslationManifestFileName(translationId: String) -> AnyPublisher<String, Error> {
+    func downloadAndCacheTranslationsFiles(translations: [TranslationModel]) -> AnyPublisher<[TranslationFilesDataModel], URLResponseError> {
         
-        guard let translation = cache.getTranslation(id: translationId) else {
-            
-            let errorDescription: String = "Failed to fetch manifest with translationId: \(translationId) because a translation object is not persisted in the realm database."
-            
-            return Fail(error: NSError.errorWithDescription(description: errorDescription))
-                .eraseToAnyPublisher()
+        let requests = translations.map {
+            self.downloadAndCacheTranslationFiles(translation: $0)
         }
         
-        return Just(translation.manifestName).setFailureType(to: Error.self)
+        return Publishers.MergeMany(requests)
+            .collect()
```

Comment:
Does `collect` wait until all the publishers in the merge have a response, and then notify subscribers only once with all the values?

Reply:
Hey @rachaelblue from my understanding the collect will gather all of the values of the upstream publishers once those have all completed and completed successfully.  It seems if any of the upstream publishers fail, then the entire thing immediately fails.

https://developer.apple.com/documentation/combine/publishers/mergemany/collect()